File: leetcode/array/3656_Determine_if_a_Simple_Graph_Exists/codes/erdos.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

int cmp_desc(const void* a, const void* b) {
    return (*(int*)b - *(int*)a);
}

bool simpleGraphExists(int* degrees, int n) {
    qsort(degrees, n, sizeof(int), cmp_desc);

    long long total = 0;
    for (int i = 0; i < n; i++) total += degrees[i];
    if (total % 2 == 1) return false;

    long long* prefix = calloc(n+1, sizeof(long long));
    for (int i = 0; i < n; i++) prefix[i+1] = prefix[i] + degrees[i];

    for (int k = 1; k <= n; k++) {
        long long left = prefix[k];
        int lo = 0, hi = n;
        while (lo < hi) {
            int mid = (lo + hi) / 2;
            if (degrees[mid] >= k) lo = mid + 1;
            else hi = mid;
        }
        int j = lo;

        long long right = 1LL * k * (k - 1);
        if (j > k) right += 1LL * k * (j - k);
        right += prefix[n] - prefix[j > k ? j : k];

        if (left > right) {
            free(prefix);
            return false;
        }
    }
    free(prefix);
    return true;
}

File: leetcode/array/3656_Determine_if_a_Simple_Graph_Exists/codes/erdos.py
from typing import List
import bisect

class Solution:
    def simpleGraphExists(self, degrees: List[int]) -> bool:
        n = len(degrees)
        degrees.sort(reverse=True)
        
        total = sum(degrees)
        if total % 2 == 1:
            return False
        
        prefix = [0] * (n + 1)
        for i in range(n):
            prefix[i+1] = prefix[i] + degrees[i]
        
        for k in range(1, n + 1):
            left = prefix[k]
            j = bisect.bisect_left(degrees, k, lo=0, hi=n, key=lambda x: -x)
            lo, hi = 0, n
            while lo < hi:
                mid = (lo + hi) // 2
                if degrees[mid] >= k:
                    lo = mid + 1
                else:
                    hi = mid
            j = lo
            
            right = k * (k - 1)
            right += k * max(0, j - k)
            right += prefix[n] - prefix[max(j, k)]
            
            if left > right:
                return False
        return True
